bin/monitor

```diff
@@ -2,62 +2,96 @@
 # frozen_string_literal: true
 
 require_relative "../loader"
+clover_freeze
 
 MONITORABLE_RESOURCE_TYPES = [PostgresServer, MinioServer, GithubRunner]
+resources = {}
+mutex = Mutex.new
 
-sessions = {}
-ssh_threads = {}
-pulse_threads = {}
-
-clover_freeze
+producer = Thread.new do
+  loop do
+    mutex.synchronize do
+      MONITORABLE_RESOURCE_TYPES.flat_map { _1.all }.each do |r|
+        resources[r.id] ||= {
+          model: r,
+          session: r.reload.init_health_monitor_session,
+          ssh_thread: nil,
+          pulse: {}
+        }
+      rescue Sequel::NoExistingObject
+      end
 
-loop do
-  resources = MONITORABLE_RESOURCE_TYPES.flat_map { _1.all }
+    sleep 5 * 60
+  end
+rescue => ex
+  Clog.emit("Resource loading is failed.") { {health_monitor_resource_load_failure: {exception: Util.exception_to_hash(ex)}} }
 
-  resources.each do |r|
-    break if Thread.list.count + 2 > Config.max_monitor_threads
-    ssh_threads[r.id] ||= Thread.new do
-      loop do
-        sessions[r.id] = r.init_health_monitor_session
+  # Since we are in a thread, raised exceptions will not be sent to the main
+  # thread, but the thread we are in will exit, causing not to be able to
+  # monitor new resources. We are exiting the whole process to restart it.
+  # Alternatively, we could just ignore the exception and retry, this keeps
+  # the same behavior as the respirate.
+  ThreadPrinter.run
+  Kernel.exit!
```

Comment:
I'm not sure if this is the correct call. Should we ignore & retry instead?

Reply:
https://github.com/ubicloud/ubicloud/pull/1340